File: backend/api/web/models/project.py
"""APIs for Project Table"""
from typing import Optional

from modules.types import PMObject


class CreateProjectBody(PMObject):
    organization_id: str
    name: str
    description: Optional[str] = None


class ProjectInstance(PMObject):
    id: int
    uuid: str
    created_at: str

    name: str
    api_key: str
    version: int
    online: bool

    organization_id: str

    description: Optional[str] = None
    cli_access_token: Optional[str] = None
    
    is_public: Optional[bool] = None

class ProjectDatasetInstance(PMObject):
    dataset_uuid: str
    dataset_name: str
    dataset_description: Optional[str] = None
    eval_metric_uuid: str
    eval_metric_name: str
    eval_metric_description: Optional[str] = None
    function_model_uuid: Optional[str] = None
    function_model_name: Optional[str] = None
    

File: frontend/src/types/Project.ts
// Interfaces for Project creation and retrieval processes

/**
 * Interface for project creation request.
 */
export interface CreateProjectRequest {
  organization_id: string;
  name: string;
  description?: string;
}

/**
 * Interface for the request to read a project's information.
 */
export interface ReadProjectRequest {
  uuid: string;
}

/**
 * Interface for the request to read an organization's projects.
 */
export interface ReadOrgProjectsRequest {
  organization_slug: string;
}

/**
 * General Project interface for representing a project in the system.
 */
export interface Project {
  id: number;
  uuid: string;
  created_at: string;
  name: string;
  api_key: string;
  version: number;
  online: boolean;
  organization_id: string;
  description: string;
  cli_access_token: string;
  is_public: boolean;
}
